src/Equations/Fluid/ComptonRateTerm.cpp

```diff
@@ -0,0 +1,109 @@
+
+/**
+ * Implementation of a class which represents the Gamma_compton contribution to the n_re equation.
+ * Employs the analytical growth rate calculated by RunawayFluid.
+ */
+using namespace DREAM;
+/**
+ * Constructor.
+ */
+ComptonRateTerm::ComptonRateTerm(
+    FVM::Grid *g, FVM::UnknownQuantityHandler *uqn,
+    RunawayFluid *rf, real_t scaleFactor
+) : EquationTerm(g), REFluid(rf), scaleFactor(scaleFactor) {
+
+    this->AllocateGamma();
+
+    this->id_n_tot   = uqn->GetUnknownID(OptionConstants::UQTY_N_TOT);
+}
+
+/**
+ * Destructor.
+ */
+ComptonRateTerm::~ComptonRateTerm() {
+    DeallocateGamma();
+}
+
+/**
+ * Allocate memory for the runaway rate.
+ */
+void ComptonRateTerm::AllocateGamma() {
+    this->gamma = new real_t[this->grid->GetNr()];
+}
+
+/**
+ * Free memory for the runaway rate.
+ */
+void ComptonRateTerm::DeallocateGamma() {
+    delete [] this->gamma;
+}
+
+/**
+ * Method called when the grid has been rebuilt.
+ */
+bool ComptonRateTerm::GridRebuilt() {
+    DeallocateGamma();
+    AllocateGamma();
+
+    return true;
+}
+
+void ComptonRateTerm::Rebuild(const real_t, const real_t, FVM::UnknownQuantityHandler *uqn) {
+    const len_t nr = this->grid->GetNr();
+
+    for (len_t ir = 0; ir < nr; ir++)
+        this->gamma[ir] = REFluid->GetComptonRunawayRate(ir);
+
+    this->data_n_tot   = uqn->GetUnknownData(id_n_tot);
+}
+ 
+/**
+ * Set jacobian block for this term. As for the avalanche growthrate, we neglect derivatives of pstar
+ * so that the compton source is treated as a constant times the total electron density
+ */
+void ComptonRateTerm::SetJacobianBlock(const len_t, const len_t derivId, FVM::Matrix *jac, const real_t*){
+    const len_t nr  = this->grid->GetNr();
+    len_t offset = 0;
+
+    for(len_t ir=0;ir<nr;ir++){
+        const len_t np1 = this->grid->GetMomentumGrid(ir)->GetNp1();
+        const len_t np2 = this->grid->GetMomentumGrid(ir)->GetNp2();
+
+        if(derivId==id_n_tot){
+            jac->SetElement(offset+np1*(np2-1)+0,offset+np1*(np2-1)+0,this->gamma[ir]/this->data_n_tot[ir]);
+        }
+        offset += np1*np2;
+    }
+}
+
+/**
+ * Set the linear operator matrix elements corresponding
+ * to this term.
+ * The compton growth rate is independent of n_re, which
+ * is the quantity this operator operates on, so we just
+ * set the rhs equal to the compton growth rate
+ */
+void ComptonRateTerm::SetMatrixElements(FVM::Matrix*, real_t *rhs) {
+    this->SetVectorElements(rhs, nullptr);
+}
+
+/**
+ * Set the non-linear function vector for this term.
+ */
+void ComptonRateTerm::SetVectorElements(real_t *vec, const real_t*) {
+    len_t offset = 0;
+    
+    for (len_t ir = 0; ir < nr; ir++) {
+        const len_t np1 = this->grid->GetMomentumGrid(ir)->GetNp1();
+        const len_t np2 = this->grid->GetMomentumGrid(ir)->GetNp2();
+
+        // Insert at p=0, xi=1
+        vec[offset + np1*(np2-1) + 0] += scaleFactor*gamma[ir];
+
+        offset += np1*np2;
+    }
+}
+
```

Comment:
I would suggest making this a term that should be added in the (id_n_re, id_n_tot) block of the eqsys in n_re.cpp. Then you get some things for free which helps make setvectorelements, jacobian etc cleaner (and lets us keep matrixelements implicit too and not use the rhs only). If you don’t feel comfortable doing it, and don’t mind, I could look into it after the merge - it’s a small change really.
We should also consider adding pstar derivatives. In the avalanche term we have an approximate expression where we assume 
pcrit ~ 1/sqrt(ntot*(E-Eceff)).
Could perhaps make that approximation here as well?

Reply:
I think I know how to do it now, I was just a bit narrow-minded when I started and for some reason wanted to keep all the fluid runaway terms on the same operator. We'll see who gets at it first ;).

Reply:
I was also thinking that we shouldn't allocate memory for gamma and store in in the equation term, since it's already stored in runawayfluid and can be retrieved when it is needed in the Set... functions.

Reply:
Regarding pstar derivatives: should we use pstar or the same pcrit that appears in the avalanche term in the compton source?

Reply:
What is your opinion? 
I want to try it out with a benchmark against a kinetic description of the compton source, and then pick the most accurate option.